### ISSUE TITLE
Great fix on planned obsolescence product but executable gives a memory error on larger than ~ 25 MB djvu files 

### ISSUE THREAD

**Issue by polbel11** [reconSuave/RollingStoner#2]
Hi, this is great news after 14 years!

 i've tried the executable on an Epyc 7282 server 2019 with 128 GB RAM and on djvu files larger than ~ 25MB it gives for example:

Processing page 81 of 81
Page size: 2437x3333
Page bitmap saved successfully, converting to RGB.
Page 81 added to PDF and temporary file removed.
Traceback (most recent call last):
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\app6.py", line 142, in <module>
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\app6.py", line 135, in main
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\app6.py", line 111, in process_file
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1065, in output
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 246, in close
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1637, in _enddoc
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1585, in _putresources
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1514, in _putimages
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1546, in _putimage
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1903, in _putstream
  File "C:\Users\ADMINI~1\AppData\Local\Temp\ONEFIL~1\fpdf\fpdf.py", line 1918, in _out
MemoryError

When it comes to programming i'm just a script kiddie but i found this link:
https://stackoverflow.com/questions/42057576/how-to-deal-with-memoryerror-in-python-code
and i have mixtral installed in faraday on this server so i could ask it how to fix in the source code but if anyone found the problem already and fixed it  i'd appreciate if he would tell me here.

To reconSuave, thanks for the great job! it's already a lot better than nothing and i'll contribute what i can. This is my first communication on github so please forgive my inexperience... my email is paul.bellehumeur @ gmail.com if someone needs to contact me. 

EDIT 2024 05 11 - Here are results of trying at least twice (as some that didn't work at first did on second try) on each djvu with the compiled exe:

Disk   pdf  --   djvu --  Success rate --  Done -- Smallest --  Largest -- Notdone --  Smallest --  Largest

2---   363.  / .405  ,,,   89.6%     .................................        7.6 MB .......   32.2 MB      ........................         24.2 MB ....  40.4 MB
3---   191  /  286   ,,,   66.8%     .................................        12.2 MB .......   35.3 MB      ........................         28.0 MB ....  69.7 MB
4---   200  /  276   ,,,  72.5%      .................................        14.7 MB .......   38.4 MB      ........................         24.4 MB ....  66.3 MB

There is a size overlap between done and not done files so individual compression ratios probably determine success rate within the 3.5 GB RAM in 32-bit x86 memory space. If running the script from python leaves more of that space free for the decryption dll it might work for the remaining undecrypted files. So I'll try this next and if that is not enough will install the software in a virtual windows XP because Vista was fresh out when the DVDs were released and it is nfg. I dont't understand why this would work with the OS taking a lot of the 3.5GB but hey worth a try!

EDIT 2024 05 13 - Per this link, 
https://forums.ni.com/t5/LabVIEW/Converting-32-bit-dll-into-64-bit/td-p/4295846
porting the bondi dll from win32 to 64-bit would require its source code but would allow to give as much memory as needed.

EDIT 2024 05 12 - In python 3.10.4 after figuring how to install pillow fpdf I ran the script in the Issues folder of dis2 on the first file  (RS0096_19711125.djvu 28553 kB)  that gave a mem error running the compiled version and i got stuck with a different memory error:

...
Processing page 81 of 81
Page size: 2437x3333
Page bitmap saved successfully, converting to RGB.
Page 81 added to PDF and temporary file removed.
Traceback (most recent call last):
  File "D:\Rolling Stone Cover to Cover - The First 40 Years\Disc 2 - Nov 1967 - Dec 1983\Bondi\Rolling Stone\Issues\bondi_to_pdf.py", line 144, in <module>
    main()
  File "D:\Rolling Stone Cover to Cover - The First 40 Years\Disc 2 - Nov 1967 - Dec 1983\Bondi\Rolling Stone\Issues\bondi_to_pdf.py", line 137, in main
    process_file(os.path.join(args.path, file), actions, output_dir)
  File "D:\Rolling Stone Cover to Cover - The First 40 Years\Disc 2 - Nov 1967 - Dec 1983\Bondi\Rolling Stone\Issues\bondi_to_pdf.py", line 113, in process_file
    pdf.output(output_pdf_path)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1065, in output
    self.close()
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 246, in close
    self._enddoc()
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1637, in _enddoc
    self._putresources()
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1585, in _putresources
    self._putimages()
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1514, in _putimages
    self._putimage(info)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1546, in _putimage
    self._putstream(info['data'])
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1903, in _putstream
    self._out(s)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python310-32\lib\site-packages\fpdf\fpdf.py", line 1918, in _out
    self.buffer+=s+"\n"
MemoryError

Gave the script to phind.com which came back with hogwash fix that was already in the original [ gc.collect() ] so it may pretend to be as good as or better than gpt4 turbo it does not find a solution as per:  

    Batch Processing: As suggested in 0, reducing the batch size might help manage memory usage more effectively. This approach can be particularly useful when dealing with large datasets or files. Although the original script does not explicitly implement batch processing, integrating this concept could involve processing files or data in smaller chunks rather than all at once.

    Using del and gc.collect() Strategically: According to 1 and 2, explicitly deleting objects with del and then calling gc.collect() can help manage memory more effectively. This strategy is particularly useful when dealing with large objects or data structures that are no longer needed. In the context of the original script, this could mean explicitly deleting objects or variables that are no longer in use, especially after processing each file or page.

    Avoiding Memory Fragmentation: 2 mentions that calling gc.collect() at the end of a loop can help avoid memory fragmentation, which in turn helps keep performance up. This is a valuable tip for scripts that process large amounts of data or files in loops. It suggests that periodically invoking garbage collection can be beneficial, especially in scenarios where memory usage spikes due to the creation and deletion of many objects.

    Consider Subprocesses for Memory-Intensive Tasks: Another suggestion from 2 is to use subprocesses for memory-intensive tasks. This approach allows the operating system to reclaim all resources used by the subprocess once it terminates, which can be particularly effective for scripts that deal with very large files or require significant memory allocation.

Given these insights, while the original script already includes a call to gc.collect(), further improvements could involve implementing batch processing, strategically using del and gc.collect() to manage memory more effectively, avoiding memory fragmentation, and considering the use of subprocesses for memory-intensive tasks. These strategies can help mitigate the risk of encountering a MemoryError when processing large files or a large number of files.

Once i supplied it the end of the script run message it came up with suggestions about more efficient pdf libraries etc:

Answer | Phind Instant Model

The error message you're encountering, MemoryError, during the execution of the script indicates that the script is running out of memory. This typically happens when the script tries to allocate more memory than is available. Given the context of your script, which involves processing and converting images to PDF, the issue likely arises from the large size of the images being processed or the cumulative effect of processing multiple large images.

The traceback points to the fpdf library's _putstream method, which is called during the PDF output process. This suggests that the issue occurs when trying to write the final PDF file, possibly due to the large size of the images being added to the PDF.

To address this issue, consider the following strategies:

    Reduce Image Size Before Adding to PDF: Before adding images to the PDF, you can resize them to a smaller dimension. This reduces the memory footprint of each image and the total memory required for all images combined.

img = img.resize((int(img.width * 0.5), int(img.height * 0.5)))  # Example: resize to half the original size

Add this resizing step right before adding the image to the PDF with pdf.image(img_file_path, 0, 0, 210, 297).

    Use a More Memory-Efficient PDF Library: If resizing images does not sufficiently reduce memory usage, consider using a more memory-efficient PDF library that can handle large images more gracefully. Libraries like reportlab or pdfkit (which uses WebKit rendering engine) might offer better performance for large documents.

    Process Images in Batches: If you're processing a large number of images, consider processing them in smaller batches. This can help manage memory usage by ensuring that only a subset of images is loaded into memory at any given time.

    Increase Available Memory: If possible, increase the available memory on your machine. This can be done by closing unnecessary applications or increasing the physical RAM in your system.

    Optimize Image Compression: When saving images to disk, ensure you're using a compression format that balances quality and file size. JPEG is a good choice for photographs, but ensure the quality setting is low enough to reduce file size.

    Explicitly Delete Objects: After processing each image and before starting the next, explicitly delete the image object from memory using del img and then call gc.collect() to force garbage collection.

Here's an example of how you might implement the first suggestion:

img = img.resize((int(img.width * 0.5), int(img.height * 0.5)))  # Resize image to half its original size
img = img.convert("RGB")  # Convert to RGB if necessary
img_file_path = os.path.join(output_dir, f"temp_img_{page}.jpg")
img.save(img_file_path, "JPEG", quality=85)  # Save with reduced quality for smaller file size
pdf.add_page()
pdf.image(img_file_path, 0, 0, 210, 297)  # Add resized image to PDF
os.remove(img_file_path)  # Clean up temporary file

Implementing these strategies should help mitigate the MemoryError by reducing the memory footprint of the images being processed and the overall memory usage of the script.
 with 
So next is making a virtual windows xp  to see if bondi app can be spied in it to extract opened issues no done with script.. Feedback to come asap.

**Comment by reconSuave** [comment]
> In python 3.10.4 after figuring how to install pillow fpdf I ran the script in the Issues folder of dis2

Hey, thanks for your detailed issue report. It sounds like the problem is that you're trying to convert the DJVU files directly from the DVD's, as I mentioned in the README this *will* result in errors and isn't supported. You should **copy the DJVU files to your hard disk first and then convert them**. The errors you're encountering are due to the DVD drive's read-buffer getting swamped, the original BondiReader software only pulled 1 image at a time into the buffer. If you still get this error when converting from your local disk, feel free to open another issue.

**Comment by reconSuave** [comment]
Resolved.

**Comment by reconSuave** [comment]
Resolved.

**Comment by polbel11** [comment]
Thanks for the reply, Nope. The files are on a pcie nvme because i don't
have the DVDs. I read carefully your instructions and even tried in a win 7
vm but cannot print to pdf from bondi.

Kind Regards,

Paul

**Comment by reconSuave** [comment]
Well, dang. I was hoping that would do it. Went ahead and reopened this issue and I'll take another look.

**Comment by polbel11** [comment]
mighty nice of You sir !

**Comment by reconSuave** [comment]
# Thinking about some possible fixes...
## MemoryError
I've been trying pretty hard to force this error and haven't been able to find it.
Since I haven't encountered this error and I haven't been able to reproduce it, at this point I'm just kind of thinking out loud.
I'll start by telling you what I know works for sure. I've tested the program in the following environment(s): 
- Windows 11 (Education edition, version 23H2). 
- Intel Core i5-8259U CPU@2.30GHz
- 48GB RAM
- Results of converting one of the largest DJVU files (104 pages, 27MB) in this environment:
- ![Image of terminal output showing successful conversion](https://i.imgur.com/O1BLbzK_d.webp?maxwidth=760&fidelity=grand)
- Also tested on Linux Wine (works quite well, which is somewhat shocking, really, as nothing has ever worked on wine! 😀) and I haven't been able to reproduce this error in that environment either (most recent version of Parrot (debian) running from another partition of the above machine)
- UPDATE: just now tested it on Windows 10 and similarly unable to recreate this error.
- **If you're using a Windows version < 10, that may well be the cause of your problem**.
> **TODO**: I'll add an explicit warning to the README that earlier Windows versions are not supported. Whatever the cause of this specific issue, I don't have access to any earlier versions of Windows and I never intended to support them. 

Some random thoughts:
- Windows will throw these kinds of generic MemoryError exceptions for a huge number of reasons, often unrelated to actual memory problems.
- You see them e.g., if Windows finds the dll's for an incompatible version of MSVC due to $PATH misconfiguration or etc.
- This program does require MSVC, and MSVC71 is the only compatible version. Since I've included the official Microsoft msvcr71.dll and msvcp71.dll redistributables with the executable package, this shouldn't be the cause of this issue. As long as you run the executable from the same working directory as the MSVC71 dll's, Windows *should* find them before it finds any other version of MSVC on $PATH (typically Windows searches first the working directory and only then resorts to searching folders from $PATH). This being said, there are probably situations where Windows would behave differently, since Windows is designed by mutant-space clowns, a deranged monster towering over its victims like a ghastly, flagellating monument to inexplicable misbehavior. ***But I digress...***

Some things I might do:
1. recompile executable with freeze support (I have no idea why this would help, but it sounded good in my head)
2. change the PDF library or add an option to choose from multiple libraries (there's nothing wrong with the PDF library, but there are countless ways to make a PDF and if it helps then great. Even so, there's not much difference between Python PDF libraries, at least with respect to factors that are likely to effect this error, they're all going to perform essentially the same operation at the step when you're seeing it throw this exception)

Some things I will do soon that might help:
1. add an option to just dump the bitmaps as PNG or JPEG into an organized directory tree. The pdf-creation step is, in some sense, completely arbitrary and chosen as a matter of convention. The functional portion of this program pulls the raw bitmap of each issue page as a PIL Image that you can do just about anything with (for example, I've built a program that serves the page-images over the web, directly from the DJVU files, kind of like a modern version of BondiReader. Check my repo's from time to time, as I'll be putting this up soon, just finishing up the templating of the web interface and adding some features, tentative timeline: ~2 weeks or so). But this might be a workaround (albeit a hackish and inelegant one) for your problem, as you seem to be encountering this error on the PDF-creating step. As long as you can dump the bitmaps to some image files that are saved in some kind of organized fashion, then you can use any number of pdf-creation tools to create a PDF from each image-directory (there are so many tools you can use for this that mentioning any would seem a bit presumptuous, but my personal favorite is [PDFCPU](https://github.com/pdfcpu/pdfcpu), it's written in Go, so it's really, really fast. Anyone who's tried to process a bunch of large PDF's will appreciate the value of this).
2. add some different ways of processing the image-output, possibly including a choice of PDF-processing libraries and maybe some options for different output formats. Actually I've always intended to do this, the machinery of this script is pretty extensible, the business-end of it happens when it decrypts the page-bitmaps, after that happens then you have the raw PIL image-data, and you can do anything with it that's possible to do with an image. So adding some options for processing the images will offer some backup strategies for achieving a successful conversion, the important part is of course accessing the content you own, adding more ways to do this will add some flexibility to work around issues like this if they pop up due to the peculiarities of any particular environment.
3. in general, there's a whole-bunch of work I've done relative to these Bondi files that I need to put on GitHub. Like I said above, there's a web-interface for reading the issues directly from the original secure DJVU files, there's also a converter for the New Yorker I'm almost done with, a collection of developer-tools for parsing and handling Bondi/FataMorgana secure DJVU, and much de-compilation/reversing of the FataMorgana library (a modified/partial implementation of Celartem's proprietary DJVU standard, **not** to be confused with the open-source DJVU standard, Celartem's DJVU is a cryptic black-box in some key ways and I'd like to at least document what I've learned on the public internet because Celartem barely exists anymore and DJVU, being a dying technology anyway, is rapidly approaching obscurity in such a ways that soon there probably won't be anybody who knows anything about it). But let me return from yet another painful digression: if nothing I've suggested above resolves this error for you, a larger collection of tools and information might help you isolate the cause of the issue or at least offer you some alternative ways to achieve a successful result.

Conclusions:
- I have no idea why you're seeing this error and I can't reproduce it.
- If you're using a version of Windows that is older than Windows 10, that may well be the cause of the issue and such cases are not supported. Try using Windows 11 or 10, if possible. Additionally, if you're running Windows on a VM, you might consider trying this on bare-metal instead (not sure why that would cause problems but my gut-instinct is that it *could*). In general, try to recreate the environment specified above, if you can, because it's proven to work.
- Sorry for not being more helpful. But keep me updated, and I'll help if I can

**Comment by polbel11** [comment]
OOPS here is the sysinfo
Microsoft Windows [Version 10.0.17763.5820]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Users\Administrator>systeminfo

Host Name:                 EPYC7282-3GPU
OS Name:                   Microsoft Windows Server 2019 Standard
OS Version:                10.0.17763 N/A Build 17763
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Server
OS Build Type:             Multiprocessor Free
Registered Owner:          Windows User
Registered Organization:
Product ID:                00429-70000-00000-AA065
Original Install Date:     1/6/2024, 12:42:36 PM
System Boot Time:          5/22/2024, 5:35:35 PM
System Manufacturer:       Supermicro
System Model:              Super Server
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: AMD64 Family 23 Model 49 Stepping 0 AuthenticAMD ~2800 Mhz
BIOS Version:              American Megatrends Inc. 2.4, 12/27/2021
Windows Directory:         C:\Windows
System Directory:          C:\Windows\system32
Boot Device:               \Device\HarddiskVolume1
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-05:00) Eastern Time (US & Canada)
Total Physical Memory:     130,935 MB
Available Physical Memory: 107,752 MB
Virtual Memory: Max Size:  131,959 MB
Virtual Memory: Available: 96,951 MB
Virtual Memory: In Use:    35,008 MB
Page File Location(s):     C:\pagefile.sys
Domain:                    WORKGROUP
Logon Server:              \\EPYC7282-3GPU
Hotfix(s):                 11 Hotfix(s) Installed.
                           [01]: KB5037933
                           [02]: KB4486153
                           [03]: KB4577586
                           [04]: KB4589208
                           [05]: KB5005112
                           [06]: KB5037765
                           [07]: KB5008287
                           [08]: KB5032306
                           [09]: KB5034863
                           [10]: KB5035963
                           [11]: KB5037017
Network Card(s):           3 NIC(s) Installed.
                           [01]: Intel(R) I210 Gigabit Network Connection
                                 Connection Name: Ethernet
                                 Status:          Media disconnected
                           [02]: Intel(R) I210 Gigabit Network Connection
                                 Connection Name: Ethernet 2
                                 DHCP Enabled:    Yes
                                 DHCP Server:     192.168.2.1
                                 IP address(es)
                                 [01]: 192.168.2.120
                                 [02]: fe80::5eab:d9b9:495b:16fd
                           [03]: VirtualBox Host-Only Ethernet Adapter
                                 Connection Name: Ethernet 3
                                 DHCP Enabled:    No
                                 IP address(es)
                                 [01]: 192.168.56.1
                                 [02]: fe80::248f:bf6a:c6af:fb68
Hyper-V Requirements:      VM Monitor Mode Extensions: Yes
                           Virtualization Enabled In Firmware: Yes
                           Second Level Address Translation: Yes
                           Data Execution Prevention Available: Yes

The drive is made of a striped array of 4 x 2TB pcie m.2 SSDs (2 x KINGSTON SNV2S2000G + INTEL SSDPEKNU020TZ + XG7000-2TB ) giving this read benchmark on crystaldiskmark:

------------------------------------------------------------------------------
CrystalDiskMark 8.0.5 x64 (C) 2007-2024 hiyohiyo
                                  Crystal Dew World: https://crystalmark.info/
------------------------------------------------------------------------------
* MB/s = 1,000,000 bytes/s [SATA/600 = 600,000,000 bytes/s]
* KB = 1000 bytes, KiB = 1024 bytes

[Read]
  SEQ    1MiB (Q=  8, T= 1): 36927.241 MB/s [  35216.6 IOPS] <   198.62 us>
  SEQ    1MiB (Q=  1, T= 1): 37409.688 MB/s [  35676.7 IOPS] <    27.89 us>
  RND    4KiB (Q= 32, T= 1):  2674.384 MB/s [ 652925.8 IOPS] <    47.38 us>
  RND    4KiB (Q=  1, T= 1):  3112.694 MB/s [ 759935.1 IOPS] <     1.21 us>

[Write]
  SEQ    1MiB (Q=  8, T= 1):     0.000 MB/s [      0.0 IOPS] <     0.00 us>
  SEQ    1MiB (Q=  1, T= 1):     0.000 MB/s [      0.0 IOPS] <     0.00 us>
  RND    4KiB (Q= 32, T= 1):     0.000 MB/s [      0.0 IOPS] <     0.00 us>
  RND    4KiB (Q=  1, T= 1):     0.000 MB/s [      0.0 IOPS] <     0.00 us>

Profile: Default
   Test: 1 GiB (x5) [D: 14% (1072/7452GiB)]
   Mode: [Admin]
   Time: Measure 5 sec / Interval 5 sec 
   Date: 2024/05/31 15:14:01
     OS: Windows Server 2019 Standard 1809 [10.0 Build 17763] (x64)
For some reason crystaldiskmark does not know how to write on it. maybe that would be the cause with the script also...
will try on a more conventional drive. Be back soon...

**Comment by reconSuave** [comment]
I seem to recall Windows Server having some kind of write-protection enabled by default as a security measure. That might be what's going on. At any rate, I've never tested this on Windows Server and don't have much experience with it, but do try it on another machine and I suspect you'll have good results.

**Comment by polbel11** [comment]
Hi, i tried on a windows 10 home notebook with the same files giving the memory errors on the larger files from disk 2. Will upgrade the machine to 11 and try to finish the decryption. Be back soon. Here is the sysinfo for the w10 machine:


C:\Disc 2 - Nov 1967 - Dec 1983\Bondi\Rolling Stone\Issues>systeminfo

Nom de l’hôte:                              WEBDIFFUSION
Nom du système d’exploitation:              Microsoft Windows 10 Famille
Version du système:                         10.0.19045 N/A version 19045
Fabricant du système d’exploitation:        Microsoft Corporation
Configuration du système d’exploitation:    Station de travail autonome
Type de version du système d’exploitation:  Multiprocessor Free
Propriétaire enregistré:                    WebDiffusion
Organisation enregistrée:
Identificateur de produit:                  00325-82042-37111-AAOEM
Date d’installation originale:              2021-05-09, 11:43:07
Heure de démarrage du système:              2024-06-01, 17:11:50
Fabricant du système:                       Acer
Modèle du système:                          Aspire A515-44G
Type du système:                            x64-based PC
Processeur(s):                              1 processeur(s) installé(s).
                                            [01] : AMD64 Family 23 Model 96 Stepping 1 AuthenticAMD ~2375 MHz
Version du BIOS:                            INSYDE Corp. V1.11, 2020-10-08
Répertoire Windows:                         C:\WINDOWS
Répertoire système:                         C:\WINDOWS\system32
Périphérique d’amorçage:                    \Device\HarddiskVolume2
Option régionale du système:                fr-ca;Français (Canada)
Paramètres régionaux d’entrée:              en-ca;Anglais (Canada)
Fuseau horaire:                             (UTC-05:00) Est (É.-U. et Canada)
Mémoire physique totale:                    7 549 Mo
Mémoire physique disponible:                4 517 Mo
Mémoire virtuelle : taille maximale:        8 765 Mo
Mémoire virtuelle : disponible:             4 721 Mo
Mémoire virtuelle : en cours d’utilisation: 4 044 Mo
Emplacements des fichiers d’échange:        C:\pagefile.sys
Domaine:                                    WORKGROUP
Serveur d’ouverture de session:             \\WEBDIFFUSION
Correctif(s):                               28 Corrections installées.
                                            [01]: KB5037587
                                            [02]: KB5037592
                                            [03]: KB4562830
                                            [04]: KB4577586
                                            [05]: KB4580325
                                            [06]: KB5000736
                                            [07]: KB5003791
                                            [08]: KB5011048
                                            [09]: KB5012170
                                            [10]: KB5015684
                                            [11]: KB5033052
                                            [12]: KB5037849
                                            [13]: KB5006753
                                            [14]: KB5007273
                                            [15]: KB5011352
                                            [16]: KB5011651
                                            [17]: KB5014032
                                            [18]: KB5014035
                                            [19]: KB5015895
                                            [20]: KB5016705
                                            [21]: KB5018506
                                            [22]: KB5020372
                                            [23]: KB5022924
                                            [24]: KB5025315
                                            [25]: KB5026879
                                            [26]: KB5037240
                                            [27]: KB5037995
                                            [28]: KB5005699
Carte(s) réseau:                            3 carte(s) réseau installée(s).
                                            [01]: Intel(R) Wi-Fi 6 AX200 160MHz
                                                  Nom de la connexion : Wi-Fi
                                                  État :                Support déconnecté
                                            [02]: Realtek PCIe GbE Family Controller
                                                  Nom de la connexion : Ethernet
                                                  DHCP activé :         Oui
                                                  Serveur DHCP :        192.168.2.1
                                                  Adresse(s) IP
                                                  [01]: 192.168.2.170
                                                  [02]: fe80::6991:ef56:5827:fe7a
                                            [03]: Bluetooth Device (Personal Area Network)
                                                  Nom de la connexion : Connexion réseau Bluetooth
                                                  État :                Support déconnecté
Configuration requise pour Hyper-V:         Extensions de mode du moniteur d’ordinateur virtuel : Oui
                                            Virtualisation activée dans le microprogramme : Oui
                                            Traduction d’adresse de second niveau : Oui
                                            Prévention de l’exécution des données disponible : Oui

**Comment by polbel11** [comment]
same result after migrating from w10 to w11.
it's ok for now. 75 % is much better than nothing anyway
and multiple thanks, let me know if you want another series
of magazines from the 1.5 million on my ebook server

Kind regard

Paul

**Comment by spedinfargo** [comment]
Jumping in because I'm seeing the same issue with the exact same stack as @polbel11.

Running in a pretty simple bare metal machine with 16gb RAM and processing the files from local folder.  Windows 11 Pro 23H2 (22631.3737).  Running the .py directly (the download of the executable flags as a virus for me).

I consistently get the error when processing a single file greater than the 25 MB:  RS0418_19840329.djvu and RS0437_0438_19841220_19850103.djvu

But here's an interesting thing:  I split the main .py into two scripts:  I left the original one intact but commented out the code that does the PDF processing and also the code that deletes the .JPG files.

I then created a very simple script that loops through all of the JPG files and does the PDF processing:
`
        pdf = FPDF()
        for file in os.listdir(args.path):
            img_file_path = os.path.join(output_dir, file)
            print(f"File: {img_file_path}")
            pdf.add_page()
            pdf.image(img_file_path, 0, 0, 210, 297)
        output_pdf_path = os.path.join(output_dir, "test.pdf")
        print(f"File: {output_pdf_path}")
        pdf.output(output_pdf_path)
`

When this code runs I do NOT get the memory error and I get a very nice PDF.  This doesn't make a lot of sense to me.  I'm going to try re-ordering things a bit in the main script and see if that has an effect:  maybe do all of the Bondi processing first and then read all of the image files at once and do the PDF processing on them.  Note to self:  change the naming of the temp image files to use three digits for page numbers so the PDF doesn't end up with the pages out of order.

I'll update the thread if I can narrow it down any further.

Python 3.8.6 by the way - script didn't work with 3.12 and I couldn't figure out how to get the 3.1 installed anymore (path issues).  I don't know a lot about Python in general, but would versions matter too much with any of this?

**Comment by polbel11** [comment]
Wow! I'll ask mixtral 8x22B to help me understand this.
Thanks for the update, It will be really nice if all the files
can be extracted.

Paul

**Comment by dougguitar** [comment]
Just wanted to chime in and report that I'm getting the same error stack trying to run the script on Windows 11 with 32 GB of physical RAM. I'm not sure if anyone came up with a reliable solution, but would love to hear about it if so.
Thanks all.

**Comment by wrewey** [comment]
For anyone having the memory issue, I updated the code in the python file to fit my needs and all of my larger files finish now (including one with 182 pages).  Instead of deleting the jpegs right away, I reduce the size a little bit and then save them all and process the PDF at the end and then delete the files (took @spedinfargo idea and expanded on it to keep it as one python file).

As an FYI, I am running it on an older Surface Book with only 8gb of RAM

Here is what I used if anyone needs it

[modified_bondi_to_pdf.txt](https://github.com/user-attachments/files/19538797/modified_bondi_to_pdf.txt)

**Comment by dougguitar** [comment]
Thanks, @wrewey! Using your modified script I was able to get the vast majority of files converted. I still got memory errors trying to process the largest djvu files, typically those over about 45MB in size. For those, I manually printed to PDF from the Bondi Reader app that came on the original Rolling Stone DVDs, which I was able to get running on Windows 11 using the v1.2 patch found out on the Wayback Machine. This task if finally done, and I have all 966 issues in PDF format. Thanks to all who have worked on this.

**Comment by llokey** [comment]
I will be the first to say that I am not a programmer.  I am a Windows 11 user and am having the same problem with files over 25 MB when using the executable.  I have tried to run the modified wrewey script but just don't know enough to run it.  I have installed fpdf and Pillow (PIL couldn't be found). I am now getting %1 is not a valid Win32 application.  In windows scripts %1 would indicate the first input which is the input directory for the djvu files which would not be a Win32 app.

Bottom line is that I am just not smart enough to run the script.  Any real help or an executable would be greatly appreciated.

**Comment by llokey** [comment]
Let me start out this comment by saying that I am not a programmer and know practically nothing about python.  I am trying to convert djvu files that are greater than 25MB using Windows 11.  I have installed the 32bit Python 3.13.5

When I run the "modified_bondi_to_pdf" script I get the following error:

FileNotFoundError: Could not find module 'C:\Users\Larry\BondiReader.DJVU.dll' (or one of its dependencies). Try using the full path with constructor syntax.

I put the file where the error said it wasn't as shown here.

PS C:\Users\Larry> dir C:\Users\Larry\BondiReader.DJVU.dll

    Directory: C:\Users\Larry

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----          4/6/2024   4:28 PM         860160 BondiReader.DJVU.dll

I even tried changing the full path to where the the script and file are both located.  Each time I get the same error.

Being dumb in this area, I'm sorta stuck.  Any help would be greatly appreciated.

Thanks - Larry

**Comment by llokey** [comment]
Thanks @wrewey and @spedinfargo  Knowing practically nothing about python nor being a programmer, I was able to get the modified script running and files converted (with help from Google).  I had to use Python 3.8.6.  I can now start working my way through all the 968 djvu files.

**Comment by gregmol123** [comment]
any news about new fixed exe ??? [modified_bondi_to_pdf.txt](https://github.com/user-attachments/files/19538797/modified_bondi_to_pdf.txt)    is not working for me windows 11 128GB RAM Intel 14Gen
